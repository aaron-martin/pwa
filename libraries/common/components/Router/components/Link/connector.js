/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { connect } from 'react-redux';
import { navigate } from '../../../../action-creators/history';

/**
 * Maps action dispatchers to the component props.
 * @param {function} dispatch The store dispatcher.
 * @return {Object} The extended component props.
 */
const mapDispatchToProps = dispatch => ({
  navigate: (action, href, historyState) => dispatch(navigate(action, href, historyState)),
});

export default connect(null, mapDispatchToProps);
