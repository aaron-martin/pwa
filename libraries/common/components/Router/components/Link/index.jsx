/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import ParsedLink from '../../helpers/parsed-link';
import getStyle from './style';

/**
 * Link component.
 * @param {Object} props Props for the component.
 * @returns {JSX}
 */
export default class Link extends PureComponent {
  static propTypes = {
    children: PropTypes.node.isRequired,
    href: PropTypes.string.isRequired,
    styles: PropTypes.shape(),
  };

  static defaultProps = {
    styles: {},
  };

  /**
   * Opens the link.
   * @param {Object} e Event
   */
  handleOpenLink = (e) => {
    e.preventDefault();

    const link = new ParsedLink(this.props.href);
    link.open();
  };

  /**
   * Renders the component.
   * @returns {XML}
   */
  render() {
    const { styles, children } = this.props;
    const style = getStyle(styles);

    return (
      // eslint-disable-next-line jsx-a11y/no-static-element-interactions
      <div role="link" className={`${style}`} onClick={this.handleOpenLink}>
        {children}
      </div>
    );
  }
}
