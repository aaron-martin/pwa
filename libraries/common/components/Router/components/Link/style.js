/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { css } from 'glamor';

export default (styles) => {
  let style = {
    padding: 0,
    margin: 0,
    border: 'none',
    textAlign: 'left',
    alignItems: 'stretch',
    cursor: 'pointer',
  };

  if (styles) {
    style = {
      ...style,
      styles,
    };
  }

  return css(style).toString();
};
