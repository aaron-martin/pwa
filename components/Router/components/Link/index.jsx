/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import connect from './connector';
import style from './style';

/**
 * Link component.
 * @param {Object} props Props for the component.
 * @returns {JSX}
 */
class Link extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    href: PropTypes.string.isRequired,
    navigate: PropTypes.func.isRequired,
    className: PropTypes.string,
    replace: PropTypes.bool,
    state: PropTypes.shape(),
  };

  static defaultProps = {
    className: '',
    replace: false,
    state: {},
  };

  /**
   * Opens the link.
   */
  handleOpenLink = () => {
    const action = this.props.replace ? 'REPLACE' : 'PUSH';
    this.props.navigate(action, this.props.href, this.props.state);
  };

  /**
   * Renders the component.
   * @returns {JSX}
   */
  render() {
    return (
      <div
        aria-hidden
        className={`${style} ${this.props.className}`}
        onClick={this.handleOpenLink}
        role="link"
      >
        {this.props.children}
      </div>
    );
  }
}

export default connect(Link);
