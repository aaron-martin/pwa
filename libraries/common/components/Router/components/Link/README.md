# Link
---

This component handles links to the new as well as the legacy (cake) system.
It uses `Library/classes/parsed-link/ParsedLink` to open the given link.
This will decide what to do with the link when clicked.

## Getting Started

```
import { Link } from 'Library/components';

<Link href="/item/123">
  <span>foo</span>
</Link>
```

## Props

### href (required)

_Type_: `string`<br>

Link to the old or legacy system.

###### Usage:

```
import { Link } from 'Library/components';

<Link href="/item/123">
  <span>foo</span>
</Link>
```

### children (required)

_Type_: `node`<br>

Children that will be rendered.

###### Usage:

```
import { Link } from 'Library/components';

<Link href="/item/123">
  <span>foo</span>
</Link>
```

### className

_Type_: `string`<br>
_Defaults_: `''`<br>

Styling for the underlying element.

###### Usage:

```
import { Link } from 'Library/components';
import style from './my-awesome.style';

<Link href="/item/123" className={style}>
  <span>foo</span>
</Link>
```

### tagName

_Type_: `string`<br>
_Defaults_: `null`<br>

The tag that should be used for the link.

###### Usage:

```
import { Link } from 'Library/components';
import style from './my-awesome.style';

<Link href="/item/123" tagName="a">
  Hello World
</Link>
```
---
