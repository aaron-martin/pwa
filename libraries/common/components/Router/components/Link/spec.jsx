/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { shallow, mount } from 'enzyme';
import Link from './index';

const mockConstructor = jest.fn();
const mockOpen = jest.fn();

jest.mock('../../helpers/parsed-link', () => (class {
  /**
   * Mocked version of the ParsedLink constructor.
   * @param {string} href Link location.
   */
  constructor(href) {
    mockConstructor(href);
  }

  /* eslint-disable class-methods-use-this */
  /**
   * Mocked version of open function.
   * @param {Object} history History object from react router
   */
  open(history) {
    mockOpen(history);
  }
  /* eslint-enable class-methods-use-this */
}));

describe('<Link />', () => {
  it('should render inner content of the link', () => {
    const wrapper = shallow(
      (
        <Link href="/">
          <span>Test</span>
        </Link>
      ),
      {
        context: {
          history: {
            push: () => {},
          },
        },
      }
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('span').length).toEqual(1);
    expect(wrapper.find('span').at(0).text()).toEqual('Test');
  });

  it('should construct parsed-link and open it when clicked', () => {
    const wrapper = mount(
      (
        <Link href="/category/123">
          <span>Test</span>
        </Link>
      ),
      {
        context: {
          history: {
            push: () => {},
          },
        },
      }
    );

    expect(wrapper).toMatchSnapshot();
    wrapper.find('div[role="link"]').at(0).simulate('click');

    expect(mockConstructor).toHaveBeenCalledWith('/category/123');
    expect(mockOpen).toHaveBeenCalled();
  });
});
