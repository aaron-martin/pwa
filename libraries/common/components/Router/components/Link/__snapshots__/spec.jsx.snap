// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> should construct parsed-link and open it when clicked 1`] = `
<Link
  className=""
  href="/category/123"
>
  <div
    className="css-c6ki83 "
    onClick={[Function]}
    role="link"
  >
    <span>
      Test
    </span>
  </div>
</Link>
`;

exports[`<Link /> should render inner content of the link 1`] = `
<div
  className="css-c6ki83 "
  onClick={[Function]}
  role="link"
>
  <span>
    Test
  </span>
</div>
`;
